### PR TITLE
attachments for tasks

### PR DESCRIPTION
**Pull request by Kasvit**
**STATUS**
READY

**DESCRIPTION**

**CHECK LIST**
- [x] specs added
- [x] specs passed
- [ ] code coverage >90%
- [ ] DB indexes added if needed
- [ ] seed.rb file is updated if needed
- [x] PR is reviewed manually again (to make sure you have 100% ready code)
- [ ] PR doesn't have conflicts with main branch
- [ ] I've checked new feature as logged in and logged out user if needed
- [x] "let!" is used in specs

### CODE REVIEW

File: app/controllers/account/tasks_controller.rb [hunk 1]
```diff
@@ -1,47 +1,54 @@
...
     resource.update_attributes(task_movement_params)
-    redirect_to account_workspace_project_path(parent.workspace_id, parent.id)
+    redirect_to account_workspace_project_path(parent.workspace_id, parent)
+  end
+
+  def delete_file_attachment
+    @project = parent
+    @task = resource
+    @task.files.find_by_id(params[:attachment_id]).purge
+    redirect_to account_project_task_path(parent, resource)
```
Comment (igorkasyanchuk) [inline app/controllers/account/tasks_controller.rb]: don't call parent, resource again here.
you already have @project

File: app/controllers/account/tasks_controller.rb [hunk 2]
```diff
@@ -1,47 +1,54 @@
...
+      redirect_to account_project_task_path(parent, @task)
     else
       render :new
     end
   end
 
   def edit
     @project = parent
-    @task = @project.tasks.find(params[:id])
+    @task = parent.tasks.find(params[:id])
```
Comment (igorkasyanchuk) [inline app/controllers/account/tasks_controller.rb]: why you calling again parent?

you have @project already

revert all changes

File: spec/controllers/account/dashboard_controller_spec.rb [hunk 3]
```diff
@@ -1,5 +1,17 @@
...
+  let!(:user) { create(:user) }
 
+  before do
+    sign_in user
+  end
+
+  context 'GET /account/dashboard' do
+      it 'should show index page' do
+        get :index
+        expect(response).to be_successful
```
Comment (igorkasyanchuk) [inline spec/controllers/account/dashboard_controller_spec.rb]: expect(response).to be_successful
        expect(response).to have_http_status(200)

doing the same 
leave only 1 line

File: spec/controllers/dashboard_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+require 'rails_helper'
+
+RSpec.describe DashboardController, type: :controller do
+
+  describe 'GET dashboard' do
+      it 'should get index page' do
+        get :index
+        expect(response).to be_successful
+        expect(response).to have_http_status(200)
```
Comment (igorkasyanchuk) [inline spec/controllers/dashboard_controller_spec.rb]: leave only 1 line

File: spec/factories/tasks.rb [hunk 5]
```diff
@@ -27,5 +27,14 @@
     description { Faker::Lorem.paragraph }
 
     project
+
+    trait :with_files do
+      after(:build) do |user|
+        2.times do |n|
+          file = ["image1.jpg", "image2.jpg", "asana.png", "text.txt", "book.pdf"].sample
+          user.files.attach(io: File.open(Rails.root.join('spec', 'factories', 'files', file)), filename: file, content_type: ['image/png', 'image/jpg', 'text/plain', 'application/pdf'])
```
Comment (igorkasyanchuk) [inline spec/factories/tasks.rb]: why you have array of content types? simplify and reduce formats or for each file put appropriate content type

File: spec/controllers/dashboard_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,11 @@
+require 'rails_helper'
+
+RSpec.describe DashboardController, type: :controller do
+
+  describe 'GET dashboard' do
+      it 'should get index page' do
```
Comment (igorkasyanchuk) [inline spec/controllers/dashboard_controller_spec.rb]: alignment

**Review by igorkasyanchuk — APPROVED** [review]

File: app/controllers/account/tasks_controller.rb [hunk 7]
```diff
@@ -25,23 +25,30 @@ def edit
...
   def move
     resource.update_attributes(task_movement_params)
-    redirect_to account_workspace_project_path(parent.workspace_id, parent.id)
+    redirect_to account_workspace_project_path(parent.workspace_id, parent)
+  end
+
+  def delete_file_attachment
+    @project = parent
+    @task = resource
+    @task.files.find_by_id(params[:attachment_id]).purge
```
Comment (romaslmd) [inline app/controllers/account/tasks_controller.rb]: avoid `::find_by_id`, prefer `::find_by(id: <id>)`

Comment (romaslmd) [reply app/controllers/account/tasks_controller.rb]: fyi, this action performs 3 separate find queries.. u could replace them w/ one

File: app/controllers/account/tasks_controller.rb [hunk 8]
```diff
@@ -25,23 +25,30 @@ def edit
...
+    redirect_to account_project_task_path(@project, @task)
```
Comment (romaslmd) [inline app/controllers/account/tasks_controller.rb:54]: don't avoid extra blank lines to visually separate logic. e.g.:
```
def delete_file_attachment
  @project = parent
  @task = resource
  @task.files.find_by(id: params[:attachment_id]).purge

  redirect_to account_project_task_path(@project, @task)
end
```

visually that blank line separates action logic from responding

File: app/views/account/tasks/_form.html.slim [hunk 9]
```diff
@@ -2,5 +2,6 @@
   = f.input_field :section, as: :hidden, value: @task.section
   = f.input :title, placeholder: 'Title', wrapper: :bulma_input, required: true
   = f.input :description, placeholder: 'Description', wrapper: :bulma_input
+  = f.input :files, input_html: { multiple: true }
   = f.submit "Save", class: "button is-link"
-  = link_to 'Back', account_workspace_project_path(@project.workspace_id, @project), class: 'button is-text'
+  = link_to 'Back', :back, class: 'button is-text'
```
Comment (romaslmd) [inline app/views/account/tasks/_form.html.slim]: not a very good idea
this button will fail on invalid form

find why

Comment (igorkasyanchuk) [reply app/views/account/tasks/_form.html.slim]: yeah ... it's possible. so put link to actual page (but check with new and edit actions)

File: spec/controllers/account/tasks_controller_spec.rb [hunk 10]
```diff
@@ -38,7 +38,7 @@
       }
 
       expect(response).to redirect_to account_workspace_project_path(workspace.to_param, project.to_param)
-      expect(project.tasks.row_order_asc.first).to eq project.tasks.order(id: :asc).second
+      expect(project.tasks.row_order_asc.first).to eq project.tasks.order(id: :asc).last
```
Comment (romaslmd) [inline spec/controllers/account/tasks_controller_spec.rb:43]: what r u testing here?

Comment (Kasvit) [reply spec/controllers/account/tasks_controller_spec.rb:43]: змінив на last через те, що раніше генерувалось 5 тасків, а тепер 2 і при руху вних записав last замість second

File: app/controllers/account/tasks_controller.rb (hunk 7 above)
Comment (Kasvit) [reply app/controllers/account/tasks_controller.rb]: I need 3 variables and use them in redirect_to

File: app/controllers/account/tasks_controller.rb [hunk 11]
```diff
@@ -25,23 +26,32 @@ def edit
...
   def move
     resource.update_attributes(task_movement_params)
-    redirect_to account_workspace_project_path(parent.workspace_id, parent.id)
+    redirect_to account_workspace_project_path(parent.workspace_id, parent)
+  end
+
+  def delete_file_attachment
+    @project = parent
+    @task = resource
+    @task.files.find_by(id: params[:attachment_id]).purge
```
Comment (igorkasyanchuk) [inline app/controllers/account/tasks_controller.rb]: change to find(params[...])

why you need find_by???

Comment (romaslmd) [inline app/controllers/account/tasks_controller.rb]: `@task.files.find_by(id: params[:attachment_id])&.purge` as `#find_by` may return `nil`

File: app/controllers/account/workspaces_controller.rb [hunk 12]
```diff
@@ -24,10 +23,9 @@ def create
   def edit
     @workspace = resource
   end
-
+
```
Comment (romaslmd) [inline app/controllers/account/workspaces_controller.rb:26]: eh?

Comment (romaslmd) [reply app/controllers/account/workspaces_controller.rb:26]: no need to commit this file as there're no changes

File: app/controllers/account/tasks_controller.rb [hunk 13]
```diff
@@ -25,23 +26,32 @@ def edit
...
+    redirect_to account_workspace_project_path(parent.workspace_id, parent)
   end
 
   def move
     resource.update_attributes(task_movement_params)
-    redirect_to account_workspace_project_path(parent.workspace_id, parent.id)
+    redirect_to account_workspace_project_path(parent.workspace_id, parent)
+  end
+
+  def delete_file_attachment
```
Comment (romaslmd) [inline app/controllers/account/tasks_controller.rb]: pls rename this method to be more like a REST

File: config/routes.rb [hunk 14]
```diff
@@ -15,6 +15,7 @@
       resources :tasks, except: [:index] do
         member do
           put :move
+          delete :delete_file_attachment
```
Comment (romaslmd) [inline config/routes.rb]: doesn't it look/sound weird? 
mb `remove_attachment`
or `delete :attachment`
or the most beautiful way - nested resource

File: spec/factories/tasks.rb [hunk 15]
```diff
@@ -27,5 +27,14 @@
     description { Faker::Lorem.paragraph }
 
     project
+
+    trait :with_files do
+      after(:build) do |user|
+        2.times do |n|
```
Comment (romaslmd) [inline spec/factories/tasks.rb]: nah.. there's a way better option - `transient`

File: spec/factories/files/text.txt [hunk 16]
```diff
@@ -0,0 +1 @@
+Hello world!
```
Comment (romaslmd) [inline spec/factories/files/text.txt]: what's this?

Comment (romaslmd) [reply spec/factories/files/text.txt]: looks like a fixture.. shouldn't it be within `spec/fixtures/**/*.*`?

File: spec/factories/tasks.rb [hunk 17]
```diff
@@ -27,5 +27,14 @@
...
+          file = ["image1.jpg", "image2.jpg", "asana.png", "text.txt", "book.pdf"].sample
```
Comment (romaslmd) [inline spec/factories/tasks.rb]: wtf?
pls create a list of existing fixtures and assign appropriate content type that matches file content and extension

Comment (romaslmd) [inline spec/factories/tasks.rb]: fixtures should be within `spec/fixtures/**/*.*`

File: spec/factories/files/text.txt (hunk 16 above)
Comment (Kasvit) [reply spec/factories/files/text.txt]: this is example file for attachments

File: spec/controllers/account/tasks_controller_spec.rb (hunk 10 above)
Comment (Kasvit) [reply spec/controllers/account/tasks_controller_spec.rb:43]: + це буде правитись у іншому ПР

File: config/routes.rb (hunk 14 above)
Comment (Kasvit) [reply config/routes.rb]: remove_attachment is good )

**Review by igorkasyanchuk — APPROVED** [review]

**Pull request merged**